### PR TITLE
fix: focus does not work on nested dialog 

### LINKED ISSUE #9329: #FocusTrap for nested DynamicDialogs still not working
Tab keys not working when i open dynamic dialog from another dynamic dialog, it stuck on close button so if i focus on control by using mouse keys and then i press tab key it focus again close button. I am using 9.1.5-lts version

### PR DESCRIPTION
**Pull request by mselerin**
### Defect Fixes  
Fix #9329 and #10472 

The fix use inject the parent `DynamicDialogComponent` inside the child dialog, disable the key listener, then restore it when the child dialog is destroyed.

Review and comments are more than welcome :-)

### CODE REVIEW

File: src/app/components/dynamicdialog/dialogservice.ts [hunk 1]
```diff
@@ -17,6 +18,15 @@ export class DialogService {
 
         this.dialogComponentRefMap.get(dialogRef).instance.childComponentType = componentType;
 
+        const parentDialog = this.injector.get(DynamicDialogComponent, null);
+        if (parentDialog) {
+            parentDialog.unbindDocumentKeydownListener();
+            dialogRef.onDestroy.pipe(
+                take(1),
+                tap(() => parentDialog.bindDocumentKeydownListener())
```
Comment (volvachev) [inline src/app/components/dynamicdialog/dialogservice.ts:26]: Hi, Why use tap instead of logic inside subscribe?

Comment (mselerin) [reply src/app/components/dynamicdialog/dialogservice.ts:26]: Just a personnal preference, based on a lot of code reviewing :

*   easier to see the chain of treatment for the observable
*   if the subscribe need to be extracted (for example to be used with an async pipe), the refactor is easy
*   the subscribe accept in fact 3 arguments (next, error and complete handler), so always using it with no argument reduce a misplacement for those handlers

I could resume this way of writing my code like 'declare ; prepare ; execute' 😅

Comment (volvachev) [reply src/app/components/dynamicdialog/dialogservice.ts:26]: The current project uses subscribe instead of `tap` more often. `Tap` also has 3 arguments)

Comment (mselerin) [reply src/app/components/dynamicdialog/dialogservice.ts:26]: OK that make sense (consistency is key) ;-)

**Comment by mselerin** [comment]
I've just update the code with a better solution : retrieve the parent `DynamicDialogComponent` inside the child `DynamicDialogComponent` directly, instead of modifying the `DialogService`.

It's cleaner since the service should not be aware of the key bindinds from the component.

**Comment by mselerin** [comment]
> Hi,
> 
> Reviewed and it seems that does not work with #10472

It seems that the fix does not work when the dialog is opened a second time... I'll check that soon !

**Comment by mselerin** [comment]
After some investigations, it seems that the `bindDocumentKeydownListener` was called twice. One time from my fix, the other time from `onAnimationStart`.

I've just add a guard to prevent a double keydown listener.

**Comment by mselerin** [comment]
Hi @cetincakiroglu, did you have time to test again this PR with #10472 ?

Does it still need some works ?

**Comment by cetincakiroglu** [comment]
Hi @mselerin 

I've just reviewed and tested your PR but it seems it still does not work. Here are the steps I follow, am I doing something wrong?

1- Click and open the first dialog
2- Click and open second dialog
3- Press "tab" to focus next input element

After the steps above, when I hit the tab after the second dialog is opened, only the close button receives focus instead of the next input element.

**Comment by mselerin** [comment]
Hi @cetincakiroglu , thanks for your answer.

I have followed the exact same step as you and for me the problem is solved :/

Here is my modified `ProductListDemo` component to test this out :

```html
@Component({
    template: `
        <p-table [value]="products" responsiveLayout="scroll" [paginator]="true" [rows]="5" [responsive]="true">
            <ng-template pTemplate="header">
                <tr>
                    <th pSortableColumn="name">Name <p-sortIcon field="vin"></p-sortIcon></th>
                    <th pSortableColumn="year">Image</th>
                    <th pSortableColumn="price">Brand <p-sortIcon field="price"></p-sortIcon></th>
                    <th pSortableColumn="inventoryStatus">Status <p-sortIcon field="inventoryStatus"></p-sortIcon></th>
                    <th style="width:4em"></th>
                </tr>
            </ng-template>
            <ng-template pTemplate="body" let-product>
                <tr>
                    <td>{{ product.name }}</td>
                    <td><img src="assets/showcase/images/demo/product/{{ product.image }}" [alt]="product.image" class="product-image" /></td>
                    <td>{{ product.price }}</td>
                    <td>
                        <span [class]="'product-badge status-' + product.inventoryStatus.toLowerCase()">{{ product.inventoryStatus }}</span>
                    </td>
                    <td>
                        <button type="button" pButton icon="pi pi-search" (click)="selectProduct(product)"></button>
                    </td>
                </tr>
            </ng-template>
        </p-table>

        <p-button (onClick)="openNestedDialog()" label="Nested dialog"></p-button>
    `,
    styles: [
        `
            .product-image {
                width: 50px;
                box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
            }
        `
    ],
    providers: [DialogService]
})
export class ProductListDemo {
    products: Product[];

    constructor(
        private productService: ProductService,
        public ref: DynamicDialogRef,
        public config: DynamicDialogConfig,
        private dialogService: DialogService
    ) {}

    ngOnInit() {
        this.productService.getProductsSmall().then((products) => (this.products = products));
    }

    selectProduct(product: Product) {
        this.ref.close(product);
    }

    openNestedDialog() {
        this.dialogService.open(ProductListDemo, {
            header: 'Nested Dialog',
            width: '50%'
        });
    }
}
```

This demo open a nested dialog with the same component (so you can open infinite nested dialog if you want).

Without my fix, pressing tab in the nested dialog only focus on the close button.

With my fix, the focus switch on the table headers and selection buttons.

**Comment by mselerin** [comment]
Hi, any news on this one ?

**Comment by mselerin** [comment]
Hi @cetincakiroglu ,

I have just synchronize my branch with master.
Are there any chances for this fix to be merge soon ?

Do you need any extra informations or help for this ?

**Comment by djay1977** [comment]
Hi @mertsincan, @cagataycivici
Would it be possible to merge this PR which works perfectly. We've been waiting for this fix for several years.
Thanks in advance

**Comment by cetincakiroglu** [comment]
Hi @mselerin,

Sorry for the latency, great catch I've merged it after a tiny refactor.
Thanks for the effort and support!

**Comment by yarikwest** [comment]
for me it still doesn't work :( (version 16.5.0)

**Pull request merged**